[WiFiLED] Fix HIGH CPU during fading during transitions

rvt opened the pull request:
<!-- DESCRIPTION -->

The original code used a while loop that tests for System.nanoTime()
This caused high CPU spikes during fading between transitions for the duration of fading time.

>         while (System.nanoTime() - startTime < nanos * 1000000) {
>             ;
>         }

In additional a few possible tread safety items are solved during updates of the targetSate to keep track of the current state just in case something went wrong.
Lastly a displosed method was added to ensure the driver can correctly be disposed by calling  executorService.awaitTermination(...) 

The fix is to use an additional Runnable using an schedule scheduleAtFixedRate that ensure the fading happens within an fading thread.

Signed-off-by: R. van Twisk <ries@vantwisk.nl> (github: rvt)

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -32,18 +39,17 @@
-    private InternalLedState currentState = new InternalLedState();
-    private InternalLedState currentTargetState = new InternalLedState();
+    private InternalLedState currentState = new InternalLedState(); // USe to not update the controller with the same value
+    private InternalLedState _currentTargetState = new InternalLedState();
```
please leave out this _

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -32,18 +39,17 @@
     private LEDStateDTO dtoState = LEDStateDTO.valueOf(0, 0, 0, 0, 0, 0, 0, 0);
-    private ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+    private ScheduledExecutorService executorService = Executors.newScheduledThreadPool(2);
+    private final Future<?> faderFuture[] = {CompletableFuture.completedFuture(null)};
```
Please put the brackets at the expected place (ie after the type)

martinvw replied:
Why is it an array anyway you only seem to use element 0

rvt replied:
I needed to pass a reference to itself to the fader routine so it can cancel itself after the nuymber of fader steps was reached. I am planning to modify it to use AtomicReference like the below. I believe that's more clear and less error prone.

> private AtomicReference<Future<?>> faderReference = new AtomicReference<>(CompletableFuture.completedFuture(null));

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,50 +159,112 @@ private void changeState(final InternalLedState newState) throws IOException {
+
+        public LEDFaderRunner(Future<?>[] faderFuture, InternalLedState fromState, InternalLedState toState, int steps, Function<InternalLedState,  Boolean> ledSender) {
+            this.faderArray = faderFuture;
+            fadeState=fromState;
```
Please format properly (using the eclipse ESH profile)

rvt replied:
Martin,

I looked around on the internet and in the source code but I have trouble finding the Eclipse ESH profile.  Can you kindly point me in the right direction? At this moment I use the default Idea formatting without formatting of wildcart imports. If it's the default from Eclipse, I can look into that and see if I can get that matching in Idea.

martinvw replied:
> Idea
Do you mean Intellij?

If you use the preconfigured Eclipse for openHAB you should just be able to call https://stackoverflow.com/a/15655127/3656774

For Intellij you can export it from the predefined Eclipse and import that in Intellij. For other IDE's it might depend.

But you could also just add spaces around the = 😉

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,50 +159,112 @@ private void changeState(final InternalLedState newState) throws IOException {
-            }, 0, TimeUnit.SECONDS);
-        }
+            } catch (NoRouteToHostException e) {
+                e.printStackTrace();
```
Do not use printStackTrace 3x

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,50 +159,112 @@ private void changeState(final InternalLedState newState) throws IOException {
-        }
+            } catch (NoRouteToHostException e) {
+                e.printStackTrace();
+                logger.warn("SocketException", e);
```
Wrong label and please provide some more context

martinvw commented:
@xylo can you take a look since I believe you contributed the orignal

martinvw commented:
Can you squash your commits and change commit message to include a `Signed-off-by` (to certify that you wrote it or otherwise have the right to pass it on as an open-source), see http://docs.openhab.org/developers/contributing/contributing.html#sign-your-work

For squashing / rebasing see:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

xylo commented:
@rvt Thanks for your PR.
@martinvw Thanks for your initial review. I'm currently on vacation but I would like to test the code when I'm back on July, 14th. At the moment I'm a bit unsure if this code
* will work with older LD382 devices which had quite some communication delay (did not need the busy wait),
* will still allow to change the fading color smoothly during a fade over
* will never lead to parallel communication (because of the thread pool size of 2 instead of 1)

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,50 +159,112 @@ private void changeState(final InternalLedState newState) throws IOException {
+                            });
+                    faderFuture[0] = executorService.scheduleAtFixedRate(sp, 0, fadeDurationInMs / fadeSteps, TimeUnit.MILLISECONDS);
+                    final Future<?> localFuture = faderFuture[0];
+                    try {
```
Is there a real reason to nest three try-catch-es

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,50 +159,112 @@ private void changeState(final InternalLedState newState) throws IOException {
+        final Function<InternalLedState, Boolean> ledSender;
+
+        public LEDFaderRunner(Future<?>[] faderFuture, InternalLedState fromState, InternalLedState toState, int steps, Function<InternalLedState,  Boolean> ledSender) {
+            this.faderArray = faderFuture;
```
Maybe you should also make this name consistent. Because the `faderFuture` sounds like a single entry.
And for consistency you could also add a `this.` on the next line :-)

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,50 +159,112 @@ private void changeState(final InternalLedState newState) throws IOException {
+                    }
                 }
+            } else {
+                thisFuture.cancel(true);
```
If I understand well this happens only in case of a miscalculation or bug (or maybe a rounding error)?

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,50 +159,112 @@ private void changeState(final InternalLedState newState) throws IOException {
+                    logger.debug("fadeState: {}", fadeState);
+                    if (!ledSender.apply(fadeState)) {
+                        logger.debug("fadeState: sending failed at step {}", fadeStepper);
+                        thisFuture.cancel(true);
```
This means there was an error fading, maybe it is better to leverage the default functionality of the scheduler to cancel the job as soon as an error is thrown, so just throw some (specific) RuntimeException and it will cancel no need for any further hacks imho.

rvt commented:
Martin,

first of all thank you for the extensive review, that's highly appreciated..

Your questions:
_will work with older LD382 devices which had quite some communication delay (did not need the - busy wait),_
I have one of the cheapo wifiled drivers from banggood (arilux), Currently using the LD382A with success
_will still allow to change the fading color smoothly during a fade over_
Essentially there shouldn't be a change in teh smoothess as I am using the scheduleAtFixedRate call that tries to guarantee execution times like initialDelay+period, then initialDelay + 2 * period, and so on.
From my eyeballing observation this seems to work fine.

- will never lead to parallel communication (because of the thread pool size of 2 instead of 1)

Parallel running of the fader thread was already blocked, but based on previous commands from Martin (why the use of array [] ..) I have refactored it to something better. I have done some tests with this aswell and the fader thread does get canceled and we are waiting for termination before scheduling a new fader thread.

- Is there a real reason to nest three try-catch-es
Not really, I took the code and simply added mine...  It's now changed to two nested try/catch. 

I have updated the PR and addresses the following changes:

- Added Signed-off-by  to PR
- Ensured formatting (Still Idea formatting but with spaces around = ) I need to download Eclipse and see if I can do such conversion...
- Removed underscore form variable name
- Do not use printStackTrace 3x: This was my bad, I cloned a copy of this driver and later realised this was modified.. Not sure what really happened... that said, they are all removed and logged correctly
- Now use a runtime error to signal stopping


The fading happens fine now. However...

bare with me.. I am new to openhab... 
Currently from a rule I have to do this : 

> 	BackLight_Color.sendCommand(new HSBType(startColor.hue, startColor.saturation, new PercentType("" + temp)))	
> 	BackLight_White.sendCommand(temp)

As you can see I send two commands because as far as I can see there is no way to set a RGBW(W2) color.
So what happens now is when I send these commands from a rule, I often get a java.net.SocketException: Socket closed I am guessing is that the wifiled cannot handle this very well and just doesn't accept the connection. As long as I change one command it works fine.

So my though is to refactor the fading wifidriver and capture all changes and when no changes are found within let's say 50ms one complete update to the client will be send.

One other thought is to use the fading capabilities of the wifiled driver itself, the ios app has an option to program it, but it keeps going... Not sure if we can stop it at the end of a cycle... I will do some research on that and see if this is possible...

xylo commented:
@rvt
I'm usually also setting the color and white channel right after each other in my rules. The current driver handles this very well I would say. A new version should be able to do it as well, since it's an essentual functionality to be able to set the color and white channels "at the same time". If the script leads to a SocketException I suspect that two threads try to communicate with the LED in parallel.

There should be also a test program for testing the fading driver. You just need to adapt the IP so that it fits your device. The program should run through without exceptions and should test the fading of colors, fading of the white channel, as well as very fast changes of the color channel. The last one should result in a flickering which is the intended.

rvt commented:
@xylo 

_If the script leads to a SocketException_ 
I agree on the socket exceptions, my version was indeed running multiple threads.

_There should be also a test program for testing the fading driver_
thank you for pointing me to the test case, I didn't realise this was provided like this. That shaved off some time testing. 

_The current driver handles this very well I would say._
I believe the fading should be (near??) just as good with the scheduleAtFixedRate method but with more relaxed CPU usage. Specially on small embedded devices, for example with 2 cores this can be important.
Note: On the PI I also noticed that fading isn't always good using the buzyLoop. On my dev machine (imac 2006) it's obviously very smooth at 20fps.

_since it's an essentual functionality to be able to set the color and white channels "at the same time"_
I fixed my routine and it allows for setting 'same time' colors.

This means for example using the test method testFrequentStateChanges() it will make it appear a bit more smooth because no new sockets are getting created and the new driver will use the same socket.

I did notice that with the current driver 'same command' doesn't work as expected.

Here is how I tested that:

I added 
on line 165 (right below the logger.debug line) the following code:
_System.out.println("New Socket");_
Set the fading time to 5000ms to easer visual inspection.

And used this in the test case:
```
System.out.println("green");
driver.setColor(HSBType.GREEN);
driver.setWhite(new PercentType(50));
System.out.println("done");
```

I get to see:
```
green
done
New Socket
New Socket
```
I first see the green led fading to full, after that the white led fading to full. As you can see, two sockets are opened for this. I would expect the fading happen at the same time. I believe this is because executorService starts to queue up color changes. Is this the intended behaviour? I would expect that with 'same' commands they are not queued up but rather set as a new target color.

My new code does fade to a new color when set and doesn't finish of previous fading. it will also keep the socket open for XXX ms after the last command and it won't turn on power for each fading command.

martinvw commented:
@xylo does this answer all your suggestions / questions / concerns?

xylo commented:
@rvt 
Well done and thank you very much for your contribution.  I tested your driver and also the old one.  Your driver works very well and indeed the old driver has currently a bug regarding immediate subsequent state changes which I was not aware of.  I'm pretty sure that it worked once, since I wrote tests especially for this feature.  However it looks like my last implementation destroyed it again.  I'm sorry for potential confusions.  Anyway, your driver has fixed this bug as well, so it's definitely an improvement in several regards.  Thank you very much!

The code looks fine for me, too.  Thanks also for splitting up the thread pool into 2 newSingleThreadScheduledExecutors.  This should prevent a parallel communication with the LED controller.

@martinvw 
All my questions and concerns have been addressed.

xylo commented:
@rvt 
Could you add your @author tag to FadingWiFiLEDDriver.java?
I don't know the openhab2 policy for adding contributors to the author list in the binding.xml, but in my opinion you can also add yourself to the binding.xml.

If @martinvw has nothing to add, you can probably squash your commits again and change the commit message to include a `Signed-off-by`.

rvt commented:
@xylo,
can you check if I did the right thing in my git squash? I have added the author tag, updated binding.xml and added and as far as I know I always added Signed-off-by.....

xylo commented:
@rvt 
Git checks it automatically. You can see the status above the text area. At the moment it says "Some checks haven’t completed yet": "Signed-off-by" needed on 5 commits.

So the squashing did seemingly not work. There are still 5 commits. For squashing the commits take a look at http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

rvt commented:
Gent's... ignore any change :) I have a fight with git and I am not winning

martinvw commented:
> Gent's... ignore any change :) I have a fight with git and I am not winning

It seems you won 👍

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -12,41 +12,56 @@
 
     public static final int DEFAULT_FADE_DURATION_IN_MS = 1000;
     public static final int DEFAULT_FADE_STEPS = 100;
+    public static final int KEEPCOMMOPENINMS = 1000; // After last fade keep communication channel open fs XXX MS
```
Normally you would use _ to separate different words in a constant: `KEEP_COMMUNICATION_OPEN_FOR_MS `

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -12,41 +12,56 @@
     private final int fadeSteps;
-    private boolean keepFading = false;
+
+    static {
```
Please remove this static block since it does not contain any code anymore

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -59,12 +74,29 @@ public void init() throws IOException {
                     s.white2);
             power = (s.state & 0x01) != 0;
-            currentTargetState = InternalLedState.fromRGBW(s.red, s.green, s.blue, s.white, s.white2);
+            currentFaderState = currentState = InternalLedState.fromRGBW(s.red, s.green, s.blue, s.white, s.white2);
```
Multiple assignments like this should better not be used, please make it more explicit:

```java
currentState = InternalLedState.fromRGBW(s.red, s.green, s.blue, s.white, s.white2);
currentFaderState = currentState;
```

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -12,41 +12,56 @@
-    private InternalLedState realTargetState = new InternalLedState();
     private LEDStateDTO dtoState = LEDStateDTO.valueOf(0, 0, 0, 0, 0, 0, 0, 0);
-    private ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+    private ScheduledExecutorService waiterFuture = Executors.newSingleThreadScheduledExecutor();
```
Why do you name them `Future`'s it is a `ScheduledExecutorService`?

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -59,12 +74,29 @@ public void init() throws IOException {
+                faderFuture.shutdownNow();
+            }
+        } catch (InterruptedException e) {
+            logger.warn("InterruptedException", e);
```
Given you are disposing you can ignore this exception, you especially don't have to `log.warn` it

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -59,12 +74,29 @@ public void init() throws IOException {
 
     @Override
+    public void dispose() {
+        waiterFuture.shutdown();
```
The dispose is also used in other cases for example when re-initializing if the configuration has changed. So after the dispose the initialize is called and the handler will be re-used. However in this case both executorservices will already be shutdown and will accept no further tasks.

An example of problems with this can be seen in this PR: https://github.com/openhab/openhab2-addons/pull/2437

rvt replied:
Martin,

we are actually not disposing the the handler but we are disposing the driver. WiFiLEDHandler will completely remove the driver in dispose and create a new one in WiFiLEDHandler.initialise.
I renamed the dispose method to shutdown for clarification and avoid further confusion.

martinvw replied:
Thanks, you are completely right 😄

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,49 +175,132 @@ private void changeState(final InternalLedState newState) throws IOException {
+     * Runnable that takes care of fading of the LED's
+     */
+    final static class LEDFaderRunner implements Runnable {
+        protected Logger logger = LoggerFactory.getLogger(LEDFaderRunner.class);
```
Should all these fields be protected or private? I would prefer private

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,49 +175,132 @@ private void changeState(final InternalLedState newState) throws IOException {
+        protected InternalLedState fromState;
+
+        // After the last update keep the communication channel open for XXX ms
+        protected long keepCommOpenForMS = 1000;
```
1) This one is always overridden from the constructor maybe leave its value away. 
2) Better choose a longer name so you don't need the documentation then it is also clear on the place where you use the field.
3) imho code becomes way more readable if order the fields in the same order as the assignments in the constructor and the order in which the parameters are passed. This makes it easy to spot how the fields are used and if there are any inconsistencies.

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,49 +175,132 @@ private void changeState(final InternalLedState newState) throws IOException {
+            this.toState = toState;
+            this.steps = steps;
+            this.ledSender = ledSender;
+            this.keepCommOpenForMS = keepCommOpenForMS;
```
Do you want to able to set this dynamic or can it just use the same constant as the parent class?

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,49 +175,132 @@ private void changeState(final InternalLedState newState) throws IOException {
+                            try {
+                                sendLEDData(fs, outputStream);
+                                currentFaderState = fs;
+                                // logger.info("Current: {} {} {} {}", fs.getR(), fs.getG(), fs.getB(), fs.getWhite());
```
Remove this or make it a trace log.

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,49 +175,132 @@ private void changeState(final InternalLedState newState) throws IOException {
+                waiterFuture.schedule(() -> {
+                    try {
+                        future.get();
+                    } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

rvt replied:
future.get() will simply wait untill LEDFaderRunner decided to stop it's execution.
This can happen for a few reasons for example during an IO Exception or when we reached the end of the fading steps or during an InterruptedException.
I will modify it to 
```
                try {
                    future.get();
                } catch (InterruptedException | ExecutionException e) {
                    // Ignored
                } catch (Exception e) {
                    logger.warn("Exception", e);
                }
```
To make it clear that these are the only two that are expected but when any other exception happens it will be logged. Not sure if this is really the best way?

martinvw replied:
Given a separate thread will by default otherwise not log its okay to catch that exceptions, I think the change above make it more clear.

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java:
```diff
@@ -143,49 +175,132 @@ private void changeState(final InternalLedState newState) throws IOException {
+            final Socket socket;
+
+            // Create and Execute a new LED Fader
+            try {
```
Do you close this socket anywhere, I think you should use the try-with-resources here:

https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

rvt replied:
Try with resources won't work because we are using a schedule at fixed rate tasks.
But you do make a good point about clarity. I have moved the opening and closing of the socket responsibility to the thread itself and added an init and shutdown method to open the sockets and power on the LEDs and a shutdown to tear down the socket.
Shutting down the thread will always be done in a waiter thread that shut's down the tread after it is finnished.

rvt commented:
Martin,

thank you for you patience and comments!
Again... fights with git and I am loosing... 
I want to fork again... see if I can make a new branch and send a compleely new PR... No idea if it can be attached to this one..

This is what 15 years of SVN does to you....

martinvw commented:
You can / should use force push to overwrite this branch, so it will just update this rsw.

rvt commented:
Gents,
I think I lost and I am terribly sorry for the mess-up... currently github says :	

> rvt  wants to merge 4 commits into openhab:master from **unknown repository**

Can we point this PR to the correct one again? 
The diff can be found here : https://github.com/rvt/openhab2-addons/commit/c15f5eb397b1144cdd88dbb90de32513fcbd0b8a

martinvw commented:
> Can we point this PR to the correct one again?

I do not have enough powers to do, I don't know if anyone does have enough powers 😄 

I check my own PR's it's also not possible for them, so just create a new one, this one is beyond repair
